File: src/main/java/com/classy/instagram/user/service/UserService.java
package com.classy.instagram.user.service;


import com.classy.instagram.user.dto.UserDto;
import com.classy.instagram.user.entity.User;
import com.classy.instagram.user.repository.UserRepository;
import java.util.ArrayList;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
@Slf4j
public class UserService {
    UserRepository userRepository;
    ModelMapper modelMapper = new ModelMapper();

    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }


    public UserDto findById(String id) {
        log.info("회원 정보 요청");
        return modelMapper.map(userRepository.findById(id), UserDto.class);
    }

    public UserDto update(UserDto userDto) {
        log.info("회원 정보 수정");
        User user = modelMapper.map(userDto, User.class);
        return modelMapper.map(userRepository.save(user), UserDto.class);
    }

    public List<UserDto> findUserList(String nickname) {
        log.info("회원 검색");
        List<User> userList = userRepository.findByNicknameContains(nickname);
        List<UserDto> userDtoList = new ArrayList<>();
        for (User user : userList) {
            userDtoList.add(modelMapper.map(user, UserDto.class));
        }
        return userDtoList;
    }
}

File: src/main/kotlin/com/classy/placelog/user/service/UserService.kt
package com.classy.placelog.user.service

import com.classy.placelog.common.storage.StorageService
import com.classy.placelog.user.dto.UserForm
import com.classy.placelog.user.entity.User
import com.classy.placelog.user.repository.UserRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserService(
    private val userRepository: UserRepository,
    private val storageService: StorageService,
) {
    @Transactional(readOnly = true)
    fun getUser(email: String): User {
        return userRepository.findById(email)
            .orElseThrow { IllegalArgumentException("User not found") }
    }

    @Transactional
    fun updateUser(userForm: UserForm): User {
        val profileImage = userForm.profile?.let { storageService.save(it) }

        val user: User = userRepository.findByIdOrNull(userForm.email)
            ?: throw IllegalArgumentException("User not found")

        user.update(
            name = userForm.name,
            nickname = userForm.nickname,
            imageUrl = profileImage,
        )

        return user
    }

    @Transactional(readOnly = true)
    fun findUser(
        keywords: String,
        pageable: Pageable,
    ): Page<User> {
        return userRepository.findByNicknameContains(keywords, pageable)
    }
}
